README
+=================================================================+
|  ____  ____  _     ___ _   _ _____                              |
| / ___||  _ \| |   |_ _| \ | | ____|  _ _                        |
| \___ \| |_) | |    | ||  \| |  _|   (_|_)                       |
|  ___) |  __/| |___ | || |\  | |___   _ _   Simple Portable      |
| |____/|_|   |_____|___|_| \_|_____| (_|_)  LISP Interpreter     |
|                                                                 |
+==================================================================

*** Operating Environment
=========================
  Primary target platforms: BSD UNIX, DEC OpenVMS
  Should be working on: GNU Linux, MS Windows, Apple MacOS X

*** Built-in functions
======================
  (1) List and dotted pair operations: CAR, CDR, CONS, LIST
      Well known actions:
      (a) (CAR (list 1 2 3 4)) returns 1
      (b) (CDR (list 1 2 3 4)) returns (2 3 4)
      (c) (CONS 1 2) returns (1 . 2) -- dotted pair creator
      (d) (LIST 'a 'b 3 4) returns (A B 3 4) -- list creator

  (2) Number operations: +, *, -, /, %, <, >, =, <=, >=
      Apply lambda operation on all arguments specified.
      For example: (% (* 1 2 3 4 (+ 2 2 1)) 20) returns 0, because
      ((1*2*3*4*(2+2+1)) mod 20) is exactly zero

  (3) Object operations: IS (compare references of two objects)

  (4) Logical AND/OR
  (5) READ(LN)/WRITE(LN) -- basic I/O
Note: (4) and (5) haven't been implemented yet

*** Built-in special forms
==========================
  (1) Well known SETQ, QUOTE, COND, LOAD, LAMBDA, SPECIAL and LET
      (a) (SETQ A 10) creates new variable atom of value 10 and returns 10
      (b) (QUOTE HELLO) is the full form for 'HELLO, this returns HELLO
          without any evaluation
      (c) (COND ((= a 1) 'A-IS-ONE)
                ((= a 2) 'A-IS-TWO)
                (T 'A-IS-NOT-ONE-OR-TWO))
          This prints A-IS-ONE it the A variable is 1, A-IS-TWO for a=2,
          otherwise A-IS-NOT-ONE-OR-TWO, because the result of all previous
          conditions was NIL.
      (d) (LOAD "path/to/some/file.lisp") will read commands in file.lisp
      (e) ((LAMBDA (X Y) (LIST 'X '= X 'AND 'Y '= Y)) 10 20) prints out
          (X = 10 AND Y = 20). This is unnamed lambda function, which can be
          "named" using SETQ:
            (SETQ MY-FUN (LAMBDA (X Y) (LIST 'X '= X 'AND 'Y '= Y)))
          and then called as: (MY-FUN 10 20) with the same result.
          The situation is similar for SPECIAL with exception that arguments
          will not be evaluated when the function is called.
      (f) (LET ((X 10) (Y 20)) (CONS X Y)) creates "local" variables X and Y,
          bound to 10 and 20 only from view of it's body -- (CONS X Y).
          It means that the result will be (10 . 20) independently
          on a "global" meaning of X and Y.

  (2) ISDEFINED -- returns NIL if the argument is an atom of type Undefined,
      otherwise returns T.

  (3) ISATOM, ISLIST, ISNUMBER -- check type of argument's value
  (4) BODY -- print the body of function
  (5) SHOW-MEM -- print a memory dump. Default is brief three-line statistics.
      When the FULL argument is supplied, it prints complete memory listing.
  (6) GETLPI -- LISP Process Informations: an equivalent to VMS-like $GETJPI
      Supported items: ATOMS_USED, LISTS_USED, NUMBERS_USED, MAXATOMS,
                       MAXLISTS, MAXNUMBERS, MODE.

*** Built-in ordinary atoms
===========================
  (1) Standard T and NIL
  (2) Special number values: INF and NaN

*** Data types
==============
  (1) Numbers (double precision)
  (2) Ordinary atoms
  (3) Lists
  (4) Strings -- in form " some string context " -- in atom table,
      returns itself, "" (an empty string) is evaluated to NIL

*** Internal options
====================
Each option is enabled just when it is set to a T value. If the option
is not defined or is defined to any non-T value, it's disabled;

  (1) LISP$LOG   -- enable logging of all %LISP-* messages; not set by default
  (2) LISP$TRACE -- print sub-results while evaluating; not set by default
  (3) LISP$GC    -- enable garbage collection; default is T

*** Friendly syntax error tracing
=================================
  Syntactic errors are traced as in the following example:
    --> (hello world `)
    %LISP-W-BADSYN, Bad syntax!
     (hello world `)
    ..............^
    -->
  it's much better than "syntax error on position 14" message
  which is typical on many compilers/interpreters/whatever...
  With this feature you can easily track similar typo.

*** Examples
============
Example scripts are startup.lisp and test.lisp.
startup.lisp is loaded automatically if it's in your current working directory
from which you are starting the LISP program.



CHANGELOG

V1.0.50 -- 04-MAY-2008 -- GNU ReadLine support added;

V1.0.76 -- 13-SEP-2009 -- GETLPI (Get LISP Process Information) support;
        -- 14-SEP-2009 -- Zero-list argument binding bug fixed;



test.lisp
;
; $Id: $
;
; Module:  test -- description
; Created: 22-DEC-2007 16:55
; Author:  tmr

; Simple arithmetic test helper
;(setq test$run ; Without lambda which is not implemented yet
;  '(list 'TEST$RUN (list test$args '= test$result) '=>
;         (cond ((= (eval test$args) test$result) 'SUCCESS)
;               (t                                'FAILURE))))

; Lambda version
(setq test$run
  (special (test result) (list test '= result '=>
    (cond ((= (eval test) (eval result)) 'SUCCESS)
          (t                             'FAILURE)))))

; AND, OR, READ, READLN, WRITE, WRITELN

; Basic arithmetic test
;(setq test$args '(+ 1 2 3 4 5)) (setq test$result 15) (eval test$run)
(test$run (+ 1 2 3 4 5) 15)

; +/-Inf and NaN test
;(setq test$args '(+ inf (- inf))) (setq test$result NaN) (eval test$run)
(test$run (+ inf (- inf)) NaN)

; Factorial, a typical recursion example
; Note that there is no lambda nor special built-in implemented,
; so we have to do it a little tricky way
;(setq fact$run '(eval
;  '(cond ((> fact$x 1) (* (+ (setq fact$x (- fact$x 1)) 1) (eval fact$run)))
;         ((show-mem) 1))))

; Using lambda definition
;(setq factorial (lambda (n)
;  (cond ((<= n 1) 1)
;        (t (* n (factorial (- n 1)))))))

; Iterative factorial definition
;(setq fact-iter (lambda (product counter n)
;  (cond ((> counter n) product)
;        (t (fact-iter (* counter product) (+ counter 1) n)))))
;
;(setq factorial (lambda (n) (fact-iter 1 1 n)))

; With memory listing in the deepest scope
(setq factorial (lambda (n)
  (cond ((> n 1) (* n (factorial (- n 1))))
        ((show-mem) 1))))

;(setq test$args '(eval fact$run))
;(setq fact$x   5) (setq test$result 120) (eval test$run)
;(setq fact$x 256) (setq test$result Inf) (eval test$run)

(test$run (factorial   5) 120)
(test$run (factorial 256) Inf)

(setq ! (lambda (x)
  (cond ((isnumber x) (factorial x))
        (T            'NOT-A-NUMBER))))

; vim: fdm=syntax:fdn=3:tw=74:ts=2:syn=lisp


LISP_Core.h
/*
 * $Id: LISP_Core.h 76 2009-09-14 11:51:40Z tmr $
 *
 * Module:  LISP_Core -- LISP Machine interfaces and structures
 * Created: 30-NOV-2007 23:54
 * Author:  tmr
 */

#ifndef _LISP_CORE_H_
#define _LISP_CORE_H_

#include <stdio.h>
#include <string.h>
#include <stdbool.h>

#define  LISP$MACH_ID   "LISP Machine V1.0 $Rev: 76 $"

#ifndef _HAS_STRL_
#define strlcpy(dst, src, size)  strncpy(dst, src, size)
#define strlcat(dst, src, size)  strncat(dst, src, size)
#endif

#ifdef _VMS_
#include <ssdef.h>
#else
#define SS$_SUCCESS   0
#endif


/******************************************************
 * LISP Definitions
 ******************************************************/

/* List, atom and number table size */
#define LISP$MACH_LSTTABLEN     0x17FF
#define LISP$MACH_ATMTABLEN     0x17FF
#define LISP$MACH_NUMTABLEN     0x17FF

/* Maximal atom name size */
#define LISP$MACH_ATMNAMELEN      64
#define LISP$MACH_INPMAXLEN     1024

/* Dealing with special numbers: +/-Infinite and Not a Number values */
#define LISP$INF         (1.0 / 0.0)                   /* +Inf def.  */
#define LISP$IS_INF(x)   ((double) (x) == (LISP$INF))  /* Is +Inf?   */
#define LISP$IS_NINF(x)  (LISP$IS_INF (-(x)))          /* Is -Inf?   */
#define LISP$NAN         (0.0 / 0.0)                   /* Not a Num. */

/* Macros for manipulating with references */
#define LISP$M_NULLREF               -1
#define LISP$M_setRef(ref,id,type)   ((ref) = (((id) << 2) | (type)))
#define LISP$M_getRefType(ref)       ((ref) & 3)
#define LISP$M_getRefId(ref)         ((ref) >> 2)
#define LISP$M_setRefType(ref,type)  (LISP$M_setRef ((ref), \
                                        LISP$M_getRefId (ref), type))
#define LISP$M_setRefId(ref,id)      (LISP$M_setRef ((ref), id, \
                                        LISP$M_getRefType (ref)))

/* Error name and message maximal lengths */
#define LISP$ERROR_NAMLEN         32
#define LISP$ERROR_MSGLEN        256

/* Error (neg.) and warning (pos.) codes; P=Parser, E=Eval, C=Core   */
#define LISP$_ERR$S_OK             0 /*  : Everything OK             */
#define LISP$_ERR$S_NOEVAL         1 /* P: No result from evaluator  */
#define LISP$_ERR$S_BADSYN         2 /* P: Bad input syntax          */
#define LISP$_ERR$S_LONGTOK        3 /* P: Too long token            */
#define LISP$_ERR$S_IVARGS         4 /* E: Invalid arguments         */
#define LISP$_ERR$S_NOTIMP         5 /* E: Not implemented yet       */
#define LISP$_ERR$S_UNDEF          6 /* E: Atom is not defined       */
#define LISP$_ERR$S_NAOFIL         6 /*  : Unable to open file       */
#define LISP$_ERR$S_ATMFUL        -1 /* C: Atom table is full        */
#define LISP$_ERR$S_LSTFUL        -2 /* C: List table is full        */
#define LISP$_ERR$S_NUMFUL        -3 /* C: Number table is full      */

/* S-Expression helpers */
/* NOTE: The first LISP used M-expressions, so if you like them,
 * feel free to change $S_PAR_OPEN to '[' and $S_PAR_CLOSE to ']' */
#define LISP$S_PAR_OPEN          '('  /* List beginning           */
#define LISP$S_PAR_CLOSE         ')'  /* End of list              */
#define LISP$S_APOSTROPHE       '\''  /* QUOTE 'syntax sugar'     */
#define LISP$S_EOS              '\0'  /* End Of Stream/String     */
#define LISP$S_EOL              '\n'  /* End Of Line              */
#define LISP$S_COMMENT           ';'  /* Comment beginning char   */
#define LISP$S_FLPDLM            '.'  /* Floating-point delimiter */
#define LISP$S_TOKEN_SIZE        256  /* Maximal token length     */

/* Option names */
#define LISP$M_OPT_TRACE "LISP$TRACE" /* Trace when evaluating    */
#define LISP$M_OPT_LOG   "LISP$LOG"   /* Log messages into a file */
#define LISP$M_OPT_GC    "LISP$GC"    /* Collect garbage          */

/* Misc */
#define LISP$ADDRFMT         "0x%04X" /* Something like '%p'       */
#define LISP$PROMPT          "--> "   /* Prompt for the first line */
#define LISP$PROMPT_ALT      "... "   /* Prompt for >1. input line */
#define LISP$PROMPT_RES      "<-- "   /* $S_write's result prompt */


/******************************************************
 * LISP Machine data structures
 ******************************************************/

/* Machine error indicator */
typedef struct {
  int  code;                        /* Error numeric code     */
  char name    [LISP$ERROR_NAMLEN]; /* Error string  code     */
  char message [LISP$ERROR_MSGLEN]; /* Full error description */
} LISP$Error;

/* Table type enumeration */
typedef enum { ListTab, AtomTab, NumberTab } LISP$TableType;

/* Reference to each LISP object in memory               */
/* Note: LISP$Ref was a structure in original design,    *
 * this was changed in V10R32 to better "typed-pointers" */
typedef int LISP$Ref;

/* List table record */
typedef struct {
  LISP$Ref car; /* Contents of Address Register   */
  LISP$Ref cdr; /* Contents of Decrement Register */
} LISP$ListRecord;

/* Atom type enumeration */
typedef enum {
  Undefined,
  Variable,        /* Ordinary atom           */
//  Number,          /* Number   atom           */
//  DottedPair,      /* Non-atomic S-Expression */
  BuiltInFun,
  BuiltInSpecForm,
  UserDefFun,
  UserDefSpecForm,
  UnnamedFun,
  UnnamedSpecForm
} LISP$AtomType;

/* Atom table record */
typedef struct {
  char          name [LISP$MACH_ATMNAMELEN];
  LISP$AtomType type;     /* Type of this AtomRecord */
  LISP$Ref value;    /* Link to list or number value */
  LISP$Ref bindList; /* Bind     list                */
//  void *        propList; /* Property list                */
} LISP$AtomRecord;

typedef enum { Interactive, Batch } LISP$Mode;

/* State of memory node -- used by GC and lst/num/atm lookup/alloc funs *
 * Let's describe the meaning of them:                                  *
 *   Free ... node can be used as storage for a newly created item      *
 *   Used ... node is already used (permanently)                        *
 *   Temp ... superposition of two previous -- node is probably used,   *
 *            but next garbage collection will make it Free             *
 *   Prot ... protected node for internal use                           */
typedef enum { Free, Used, Temp, Prot } LISP$MemState;

/* LISP Machine Instance (LMI) */
typedef struct {
  LISP$Error      error;                        /* LMI error indicator  */

  FILE *          logStream;                    /* Log file stream      */
  FILE *          inputStream;                  /* S-expr input stream  */
  FILE *          outputStream;                 /* S-expr output stream */
#ifdef _HAS_RDLN_
  char *          readLineBuffer;               /* ReadLine's inp. buf. */
#endif
  char            inputBuffer [LISP$MACH_INPMAXLEN]; /* Recently read   */
  int             inputReadPosition;            /* Current getChar pos. */
  int             inputErrorPosition;           /* Incorrect input pos. */

  bool            isReady;                      /* Is initialized?      */
  bool            is2ShutDown;                  /* Is LMI to be halted? */
  bool            is2Process;                   /* Is there st2be read? */
  bool            sexprEnd;                     /* S-expr read "lock"   */

  LISP$Mode       mode;                         /* Mode of processing   */

  LISP$MemState   lstIdx [LISP$MACH_LSTTABLEN]; /* Used lists   index   */
  LISP$MemState   atmIdx [LISP$MACH_ATMTABLEN]; /* Used atom    index   */
  LISP$MemState   numIdx [LISP$MACH_NUMTABLEN]; /* Used numbers index   */

  LISP$ListRecord lstTab [LISP$MACH_LSTTABLEN]; /* List table           */
  LISP$AtomRecord atmTab [LISP$MACH_ATMTABLEN]; /* Atom table           */
  double          numTab [LISP$MACH_NUMTABLEN]; /* Number table         */

  LISP$Ref        QUOTE;                        /* QUOTE reference      */
  LISP$Ref        NIL;                          /* NIL   reference      */
  LISP$Ref        T;                            /* T     reference      */

} LISP$MachIns;


/******************************************************
 * LISP Machine control operations
 ******************************************************/

/* Initialize a new LMI */
LISP$MachIns * LISP$M_init (char * name, FILE * input, FILE * output);

void LISP$M_setStreams (LISP$MachIns * lmi, FILE * input, FILE * output);

/* Destroy specified LMI */
void LISP$M_destroy (LISP$MachIns * lmi);

/* Current stream has been processed up to its end */
void LISP$M_setProcessed (LISP$MachIns * lmi);

/* Check if there are streams to be read */
bool LISP$M_is2Process (LISP$MachIns * lmi);

/* Check if the specified LMI is waiting to be halted */
bool LISP$M_is2ShutDown (LISP$MachIns * lmi);

/* Print message to both log and output stream */
void LISP$M_throwMessage (LISP$MachIns * lmi, const char severity,
                          const char * name, const char * msg, ...);

/* Check for error state */
bool LISP$M_isError (LISP$MachIns * lmi);

/* Report error to the 'output' stream */
void LISP$M_reportError (LISP$MachIns * lmi);

/* Set error */
void LISP$M_setError (LISP$MachIns * lmi, const int code,
                      const char * name, const char * msg, ...);

/* Find a first free node in the specified storage table */
int LISP$M_findFreeNode (LISP$MemState idx [], int length);

/* Allocate a new item in the list area */
LISP$Ref LISP$M_createList (LISP$MachIns * lmi);

/* Settle a new number record in the number table */
LISP$Ref LISP$M_getCreateNumber (LISP$MachIns * lmi, double num);

/* Lookup for specified atom or create a new one */
LISP$Ref LISP$M_getCreateAtom (LISP$MachIns * lmi, char * atomName);

/* Make lookup for the symbol 'option' and check if it's set to T */
bool LISP$M_isOptionEnabled (LISP$MachIns * lmi, const char * option);

/* Return true if the 'atom' is built-in, user def. or unnamed special form */
bool LISP$M_isSpecialForm (LISP$AtomRecord atom);

/* Return true if the 'atom' is built-in, user def. or unnamed function */
bool LISP$M_isFunction (LISP$AtomRecord atom);

/* Check built-in atom's name and argument requirements               *
 * NOTE: to prevent checking argcMin and/or argcMax, set them to -Inf */
bool LISP$M_checkBuiltIn (LISP$MachIns * lmi, const char * name,
                          LISP$AtomRecord bultIn,
                          LISP$ListRecord argv, int argc,
                          int argcMin, double argcMax);

/* Set the 'mark' flag on a memory node specified by the 'ref' */
void LISP$M_markMemNode (LISP$MachIns * lmi, LISP$Ref ref,
                         LISP$MemState mark, bool recurse);

/* Check the specified memory node for a 'mark' memory state */
bool LISP$M_checkMemNodeState (LISP$MachIns * lmi, LISP$Ref ref,
                               LISP$MemState mark);

/* Collect and free previously marked garbage */
void LISP$M_collectGarbage (LISP$MachIns * lmi);

/* Print dump of all memory storages */
void LISP$M_printMemoryDump (LISP$MachIns * lmi, bool full);

/* Implementation of built-in functions and special forms */
LISP$Ref LISP$M_builtInCAR (LISP$MachIns * lmi, LISP$Ref args);
LISP$Ref LISP$M_builtInCDR (LISP$MachIns * lmi, LISP$Ref args);
LISP$Ref LISP$M_builtInSETQ (LISP$MachIns * lmi, LISP$Ref args,
                               int level);
LISP$Ref LISP$M_builtInCONS (LISP$MachIns * lmi, LISP$Ref args);
LISP$Ref LISP$M_builtInArithmetic (LISP$MachIns * lmi, LISP$Ref args,
                                     int argc, const char * oper);
LISP$Ref LISP$M_builtInIS (LISP$MachIns * lmi, LISP$Ref args);
LISP$Ref LISP$M_builtInCOND (LISP$MachIns * lmi, LISP$Ref args,
                               int level);
LISP$Ref LISP$M_builtInLOAD (LISP$MachIns * lmi, LISP$Ref args,
                               int level);
LISP$Ref LISP$M_builtInSHOWMEM (LISP$MachIns * lmi, LISP$Ref args,
                                  int argc);
LISP$Ref LISP$M_builtInGETLPI (LISP$MachIns * lmi, LISP$Ref args,
                               int argc);
LISP$Ref LISP$M_builtInLET (LISP$MachIns * lmi, LISP$Ref args, int level);
LISP$Ref LISP$M_builtInISDEFINED (LISP$MachIns * lmi, LISP$Ref args);
LISP$Ref LISP$M_builtInISATOM (LISP$MachIns * lmi, LISP$Ref args);
LISP$Ref LISP$M_builtInISLIST (LISP$MachIns * lmi, LISP$Ref args);
LISP$Ref LISP$M_builtInLAMBDASPECIAL (LISP$MachIns * lmi, LISP$Ref args,
                                        bool isLambda);

/* Stack operations on list -- useful for example for binding values */
void LISP$M_listPush (LISP$MachIns * lmi, LISP$Ref* list, LISP$Ref item);
LISP$Ref LISP$M_listPop (LISP$MachIns * lmi, LISP$Ref* list);

/* Value binding functions */
void LISP$M_bind (LISP$MachIns * lmi, LISP$Ref atm, LISP$Ref val);
void LISP$M_unBind (LISP$MachIns * lmi, LISP$Ref atm);


/******************************************************
 * S-Expression processing
 ******************************************************/

/* fgetc wrapper with history feature similar to libreadline */
char LISP$S_getChar (LISP$MachIns * lmi);

/* Append char 'c' into the 'token' buffer ans optionally upcase it */
bool LISP$S_buildToken (LISP$MachIns * lmi, char ** token,
                        int i, char c, bool isCaseSensitive);

/* Check if the char 'c' is allowed to be part of atom's name */
bool LISP$S_isValidAtomNameChar (char c, int i);

/* Print a fine prompt */
void LISP$S_makePrompt (LISP$MachIns * lmi, bool primary, bool is2ReadLine);

/* Read token from an input stream and store it into the 'token' buffer; *
 * if the token is a number, set 'isNumeric' to 'true'                   */
char LISP$S_getToken (LISP$MachIns * lmi, char ** token, bool * isNumeric);

/* Helper function for evaluation process tracing */
void LISP$S_trace (LISP$MachIns * lmi, LISP$Ref res, int level);

/* Read S-Expression from the 'input' stream */
LISP$Ref LISP$S_read (LISP$MachIns * lmi);

/* $S_read's internal action */
LISP$Ref LISP$S_readAction (LISP$MachIns * lmi, int * level);

/* Evaluate result of $S_Read call */
LISP$Ref LISP$S_eval (LISP$MachIns * lmi,
                           LISP$Ref read);

/* $S_eval's internal action */
LISP$Ref LISP$S_evalAction (LISP$MachIns * lmi, LISP$Ref read,
                              int level);

/* Write evaluated result to the 'output' stream */
void LISP$S_write (LISP$MachIns * lmi,
                   LISP$Ref evaluated);

/* $S_write's internal action */
void LISP$S_writeAction (LISP$MachIns * lmi, LISP$Ref evaluated,
                         bool isTopLevel);

/* Get human readable string for $AtomType argument 'type' */
char * LISP$S_getFormatAtomType (LISP$AtomType type);

/* Load a script file and returns boolean success */
bool LISP$S_loadFile (LISP$MachIns * lmi, char * fileName);

#endif

// vim: fdm=syntax:fdn=1:tw=74:ts=2:syn=c


LISP_Main.c
/*
 * $Id: LISP_Main.c 35 2008-01-08 21:37:42Z tmr $
 *
 * Module:  LISP_Main -- LISP Interpreter's main interaction loop
 * Created: 30-NOV-2007 20:55
 * Author:  tmr
 */

#include <stdio.h>

#include "LISP_Core.h"

#ifdef _VMS_
#include <ssdef.h>
#else
/* Normal completition */
#define SS$_NORMAL        0
#define SS$_INSFARG      10
#define SS$_ACCVIO       20
#endif

#define LISP$INPUT    stdin
#define LISP$OUTPUT   stdout


int main (int argc, char * argv []) {

  int i;
  LISP$MachIns * lmi;
  LISP$Ref       ref;

  /* Initialize LISP Machine */
  lmi = LISP$M_init (argv [0], LISP$INPUT, LISP$OUTPUT);

  /* Batch mode -- load startup files */
  for (i = 0; i < argc; i++)
    LISP$S_loadFile (lmi, (!i)? "startup.lisp" : argv [i]);

  while (LISP$M_is2Process (lmi)) {
    ref = LISP$S_read (lmi);
    ref = LISP$S_eval (lmi, ref);

    LISP$S_write (lmi, ref);
  }

  LISP$M_destroy (lmi);

  return SS$_NORMAL;
}

// vim: fdm=syntax:fdn=1:tw=74:ts=2:syn=c


tmp/TRMIO.c
/*
 * $Id: $
 *
 * Module:  key -- description
 * Created: 04-MAY-2008 00:08
 * Author:  tmr
 */

#include <stdio.h>
#include <termios.h>

#include "TRMIO.h"

TRMIO$Instance * TRMIO$P_init (FILE * input, FILE * output) {

  /* Create 'instance' */
  TRMIO$Instance * trmio = (TRMIO$Instance *) malloc (sizeof (TRMIO$Instance));

  /* Set I/O streams */
  TRMIO$P_setInput (trmio, input);
  TRMIO$P_setOutput (trmio, output);

  return (trmio);
}

void TRMIO$P_destroy (TRMIO$Instance * trmio) { free (trmio); }

void TRMIO$P_setInput (TRMIO$Instance * trmio, FILE * inStream) {

  /* Flush old stream */
  if (trmio->inputStream != NULL) fflush (trmio->inputStream);

  /* Set a new stream */
  trmio->inputStream = inStream;
}

void TRMIO$P_setOutput (TRMIO$Instance * trmio, FILE * outStream) {

  /* Flush old stream */
  if (trmio->outputStream != NULL) fflush (trmio->outputStream);

  /* Set a new stream */
  trmio->outputStream = outStream;
}

int TRMIO$P_getc (TRMIO$Instance * trmio) {

  struct termios oldt,
                 newt;
  int            c, fd;

  fd = fileno (trmio->inputStream);

  tcgetattr (fd, &oldt);
  newt = oldt;
  newt.c_lflag &= ~(ICANON | ECHO);

  tcsetattr (fd, TCSANOW, &newt);
  c = fgetc (trmio->inputStream);

  tcsetattr (fd, TCSANOW, &oldt);

  return (c);
}

int main () {

  int c;
  TRMIO$Instance * trmio;

  trmio = TRMIO$P_init (stdin, stdout);

  while (c != EOF) {
    c = TRMIO$P_getc (trmio);
    printf (">>>%c<<< >>>%d<<< >>>%X<<<\n", c, c, c);
  }

  TRMIO$P_destroy (trmio);

  return (0);
}

// vim: fdm=syntax:fdn=1:tw=74:ts=2:syn=c


startup.lisp
;
; $Id: $
;
; Module:  startup -- The LISP startup script
; Created: 22-DEC-2007 16:55
; Author:  tmr
;
;  _     _____ ___________  ___  ___           _     _
; | |   |_   _/  ___| ___ \ |  \/  |          | |   (_)
; | |     | | \ `--.| |_/ / | .  . | __ _  ___| |__  _ _ __   ___
; | |     | |  `--. \  __/  | |\/| |/ _` |/ __| '_ \| | '_ \ / _ \
; | |_____| |_/\__/ / |     | |  | | (_| | (__| | | | | | | |  __/
; \_____/\___/\____/\_|     \_|  |_/\__,_|\___|_| |_|_|_| |_|\___|
;
;                          +-+-+-+-+ +-+ +-+-+-+-+-+-+-+ +-+-+-+-+
;                          |J|U|S|T| |A| |C|O|M|M|E|N|T| |T|E|S|T|
;                          +-+-+-+-+ +-+ +-+-+-+-+-+-+-+ +-+-+-+-+

; Enable garbage collector, evaluator tracing and logging
(setq LISP$GC    T) ; this is default
(setq LISP$LOG   NIL)
(setq LISP$TRACE NIL)

(setq report-error (lambda (code msg) (list '%ERROR-REPORT: code '=> msg) ))

; LISP Startup file
;(setq a 1)  ; set A=1
;(setq b 2)  ; set B=2
;(setq c 3)  ; set C=3

;(setq d (list a b c))
;(setq e (list 'a a 'b b 'c c 'd d))

;(setq f
;  (list
;    (list 'a '=> a)
;    (list 'b '=> b)
;    (list 'c '=> c)
;    (list 'd '=> d)
;    (list 'e '=> e)
;  )
;)

;(setq g
;  (list
;    (cons 'list123 (list 1 2 3))
;    (cons 'list456 (list 4 5 6))
;    (cons 'list789 (list 7 8 9))
;  )
;)

;(setq cond4a '(cond ((= a inf) 'A=INF) (t 'A<INF)))
;(setq cond4all '(cond ((<= a inf) (eval cond4a))
;                      ((<= b inf) 'B<=INF)
;                      (t          'DEFAULT)))
;(eval cond4all)   ; let's evaluate it

;(load "test.lisp")

; ---------------------- BegOfTest

; Simple arithmetic test helper
;(setq test$run ; Without lambda which is not implemented yet
;  '(list 'TEST$RUN (list test$args '= test$result) '=>
;         (cond ((= (eval test$args) test$result) 'SUCCESS)
;               (t                                'FAILURE))))

; Lambda version
(setq test$run
  (special (test result) (list test '= result '=>
    (cond ((= (eval test) (eval result)) 'SUCCESS)
          (t                             'FAILURE)))))

; AND, OR, READ, READLN, WRITE, WRITELN

; Basic arithmetic test
;(setq test$args '(+ 1 2 3 4 5)) (setq test$result 15) (eval test$run)
(test$run (+ 1 2 3 4 5) 15)

; +/-Inf and NaN test
;(setq test$args '(+ inf (- inf))) (setq test$result NaN) (eval test$run)
(test$run (+ inf (- inf)) NaN)

; Factorial, a typical recursion example
; Note that there is no lambda nor special built-in implemented,
; so we have to do it a little tricky way
;(setq fact$run '(eval
;  '(cond ((> fact$x 1) (* (+ (setq fact$x (- fact$x 1)) 1) (eval fact$run)))
;         ((show-mem) 1))))

; Using lambda definition
;(setq factorial (lambda (n)
;  (cond ((<= n 1) 1)
;        (t (* n (factorial (- n 1)))))))

; Iterative factorial definition
;(setq fact-iter (lambda (product counter n)
;  (cond ((> counter n) product)
;        (t (fact-iter (* counter product) (+ counter 1) n)))))
;
;(setq factorial (lambda (n) (fact-iter 1 1 n)))

; With memory listing in the deepest scope
(setq factorial (lambda (n)
  (cond ((> n 1) (* n (factorial (- n 1))))
        ((show-mem) 1))))

;(setq test$args '(eval fact$run))
;(setq fact$x   5) (setq test$result 120) (eval test$run)
;(setq fact$x 256) (setq test$result Inf) (eval test$run)

(test$run (factorial   5) 120)
(test$run (factorial 256) Inf)

(setq ! (lambda (x)
  (cond ((isnumber x) (factorial x))
        (T            'NOT-A-NUMBER))))

; ---------------------- EndOfTest

;(setq mycons (lambda (a b) (cons a b)))

(setq not (lambda (x)
  (cond ((is x nil)   t)
        (t          nil))))

(setq abs (lambda (x)
  (cond ((not (isnumber x)) 'NOT-A-NUMBER)
        ((>= x 0) x)
        ((<  x 0) (- x)))))

(setq good-enough (lambda (guess x)
  (< (abs (- (* guess guess) x)) 0.001)))

(setq improve (lambda (guess x)
  (/ (+ guess (/ x guess)) 2.0)))

(setq sqrt-iter (lambda (guess x)
  (cond ((good-enough guess x) guess)
        (t (sqrt-iter (improve guess x) x)))))

(setq sqrt (lambda (x)
  (sqrt-iter (/ x 2.0) x)))

(setq set (special (x y)
  (eval (cons 'setq (cons (eval x) (cons y nil))))))

(setq append (lambda (x y)
  (cond ((is x nil) y)
        ((isatom x) (cons x y))
        (t (cons (car x) (append (cdr x) y))))))

(setq reverse (lambda (x)
  (cond ((isatom x) x)
        (t (append (reverse (cdr x)) (cons (car x) nil))))))

(setq apply (special (f x) (eval (cons f x))))

(setq merge (lambda (v l)
  (cond ((is nil l)    (cons v l))
        ((<= v (car l)) (cons v l))
        (t (cons (car l) (merge v (cdr l)))))))

(setq sort (lambda (x) 
   (cond ((is nil x) x)
         (t (merge (car x) (sort (cdr x)))))))

(setq member (lambda (a s)
  (cond ((is s nil)     nil)
        ((is a (car s)) t)
        (t (member a (cdr s))))))

(setq exp (lambda (a x)
            (cond
              ( (> x 0) (* a (exp a (- x 1))) )
              ( (< x 1) 1 ) )))

(setq bool2bin (lambda (x)
  (cond ( (is x t) 1 )
        ( (is x nil) 0 )
        ( t (report-error 'BOOL2BIN-E-IVARGS "Only T/NIL arguments are accepted!") ) )))

(setq bin2bool (lambda (x)
  (cond ( (= x 1) t )
        ( (= x 0) nil)
        ( t (report-error 'BIN2BOOL-E-IVARGS "Only 0/1 arguments are accepted!") ) )))

; -----------------------------------------------------
; Logical operators -- implemented in a LISP itself,
; since their built-in friends are not implemented yet..
; Note: they are only binary (arity of two) since
; we don't support a variable lenght argument list for
; lambdas...

; Binary AND operation
(setq && (lambda (x y)
  (bin2bool (* (bool2bin x) (bool2bin y))) ))

; Binary OR operation
(setq || (lambda (x y)
  (let ( (res (+ (bool2bin x) (bool2bin y))) )
       (cond ((> res 0) t) (t nil)) )))

; A bit of set operations
(setq is-in-set (lambda (key set)
  (cond ( (not (isatom key))   (report-error 'IS-IN-SET-E-IVARGS "First argument is not an atom!") )
        ( (is set nil)         nil )
        ( (not (islist   set)) (report-error 'IS-IN-SET-E-IVARGS "Second argument is not a list!") )
        ( (is key (car set))   t )
        ( t                    (is-in-set key (cdr set)) ) )))

(setq append-unique (lambda (x y)
  (cond ( (not (isatom x)) (report-error 'APPEND-UNIQUE-E-IVARGS "First argument is not an atom!") )
        ( (is y nil)       (list x) )
        ( (not (islist y)) (report-error 'APPEND-UNIQUE-E-IVARGS "Second argument is not a list!") )
        ( (not (is-in-set x y)) (append x y) )
        ( t y ) )))

(setq conjunct (lambda (x y)
  (cond ( (is x nil) y )
        ( (is y nil) x )
        ( t (append-unique (car x) (conjunct (cdr x) y)) ) )))

(setq remove (lambda (x y)
  (cond ( (not (isatom x)) (report-error 'REMOVE-E-IVARGS "First argument is not an atom!") )
        ( (is y nil)       nil )
        ( (not (islist y)) (report-error 'REMOVE-E-IVARGS "Second argument is not a list!") )
        ( (is x (car y))   (cdr y) )
        ( t                (cons (car y) (remove x (cdr y))) ) )))

(setq disjunct (lambda (x y)
  (cond ( (is x nil) y )
        ( (is y nil) x )
        ( (not (is-in-set (car x) y)) (cons (car x) (disjunct (cdr x) y)) )
        ( t (disjunct (cdr x) (remove (car x) y)) ) )))

(setq intersect (lambda (x y)
  (cond ( (is x nil) nil )
        ( (is y nil) nil )
        ( (is-in-set (car x) y) (cons (car x) (intersect (cdr x) y)) )
        ( t (intersect (cdr x) y) ) )))

(setq conjunct-sorted
      (lambda (x y)
        (sort (conjunct x y)) ))

(setq disjunct-sorted
      (lambda (x y)
        (sort (disjunct x y)) ))

(setq intersect-sorted
      (lambda (x y)
        (sort (intersect x y)) ))

(load "mode.lisp")

; End of startup
; vim: fdm=syntax:fdn=3:tw=74:ts=2:syn=lisp
